### PR TITLE
SPOI-8925: Add example for CSVParser Operator

### PR DESCRIPTION
**Pull request by DT-Priyanka**
@shubham-pathak22  please review the application for csv parsing

### CODE REVIEW

**Comment by amberarrow** [comment]
Please add a unit test that actually verifies the result of parsing (the existing jsonparser also needs a similar test; we can create a separate JIRA for that if you prefer or just do it as part of this PR).
See ongoing discussion at: https://github.com/DataTorrent/examples/pull/16

**Comment by amberarrow** [comment]
The README.md needs a description of this new example.

File: tutorials/parser/src/main/resources/META-INF/csvParserProperties.xml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0"?>
```
Comment (amberarrow) [inline tutorials/parser/src/main/resources/META-INF/csvParserProperties.xml]: Will this properties file actually take effect when run on a cluster ? Shouldn't the file be named
properties-csvParserProperties.xml ?

Please test all examples both in local mode as well as on a cluster -- we want the user experience to be smooth and painless when running these.

File: tutorials/parser/src/main/java/com/datatorrent/tutorial/csvparser/AddDataGenerator.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.commons.lang.RandomStringUtils;
+import org.apache.commons.lang.math.RandomUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datatorrent.api.Context.OperatorContext;
+import com.datatorrent.api.DefaultOutputPort;
+import com.datatorrent.api.InputOperator;
+
+public class AddDataGenerator implements InputOperator
```
Comment (shubham-pathak22) [inline tutorials/parser/src/main/java/com/datatorrent/tutorial/csvparser/AddDataGenerator.java]: You may rename it to AdDataGenerator

File: tutorials/parser/src/main/java/com/datatorrent/tutorial/csvparser/csvParserApplication.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  public void populateDAG(DAG dag, Configuration conf)
+  {
+    AddDataGenerator dataGenerator = dag.addOperator("dataGenerator", new AddDataGenerator());
+    CsvParser parserOperator = dag.addOperator("csvParser", new CsvParser());
+    ConsoleOutputOperator consoleOperator = dag.addOperator("dataconsole", new ConsoleOutputOperator());
+    ConsoleOutputOperator errorConsoleOperator = dag.addOperator("errorConsole", new ConsoleOutputOperator());
+
+    dag.addStream("inputData", dataGenerator.out, parserOperator.in);
+    dag.addStream("parsedData", parserOperator.parsedOutput, consoleOperator.input);
+    dag.addStream("errorData", parserOperator.err, errorConsoleOperator.input);
```
Comment (shubham-pathak22) [inline tutorials/parser/src/main/java/com/datatorrent/tutorial/csvparser/csvParserApplication.java]: I would suggest we add one more console operator that would receive POJOs. Most downstream operators expect a POJO hence having it in example would be good.

File: tutorials/parser/src/main/java/com/datatorrent/tutorial/csvparser/AddDataGenerator.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+{
+  private static Logger LOG = LoggerFactory.getLogger(AddDataGenerator.class);
+  private int tuplesPerWindow;
+  private int tuplesEmitted = 0;
+  public DefaultOutputPort<byte[]> out = new DefaultOutputPort<>();
+
+  private SimpleDateFormat startDateFormat;
```
Comment (shubham-pathak22) [inline tutorials/parser/src/main/java/com/datatorrent/tutorial/csvparser/AddDataGenerator.java]: This can be transient

**Comment by DT-Priyanka** [comment]
@shubham-pathak22  addressed your review comments.

**Comment by DT-Priyanka** [comment]
@shubham-pathak22  updated as per review comments.

**Comment by DT-Priyanka** [comment]
@sandeepdeshmukh  can you please take a look and merge as @shubham-pathak22 is traveling.

**Comment by DT-Priyanka** [comment]
@shubham-pathak22  please merge if there are no more comments.

**Review by shubham-pathak22 — COMMENTED** [review]
Please update README.md as @amberarrow have suggested. Rest all looks good to me . If there are no further comments, will merge the change.

**Comment by DT-Priyanka** [comment]
Strange, I remember I had updated it for sure, but now I only can't find that edit. Anyway will write that again.

File: tutorials/parser/src/main/java/com/datatorrent/tutorial/csvparser/Campaign.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  private String campaignName;
+  private String parentCampaign;
+  private double campaignBudget;
+  private double bidPrice;
+  private int campaignId;
+  private long securityCode;
+  private boolean weatherTargeting;
+  private boolean active;
+  private boolean optimized;
+  @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
```
Comment (shubham-pathak22) [inline tutorials/parser/src/main/java/com/datatorrent/tutorial/csvparser/Campaign.java]: Why do we need this ?

Comment (DT-Priyanka) [inline tutorials/parser/src/main/java/com/datatorrent/tutorial/csvparser/Campaign.java]: We have schema restriction to look for this date format. Please check "schema" in properties file.

Comment (shubham-pathak22) [inline tutorials/parser/src/main/java/com/datatorrent/tutorial/csvparser/Campaign.java]: The annotation is not needed. Date format is read from schema file.

Comment (DT-Priyanka) [inline tutorials/parser/src/main/java/com/datatorrent/tutorial/csvparser/Campaign.java]: done

**Comment by tweise** [comment]
@amberarrow test for the parser operator should be separate JIRA for Malhar. The test for this application should ensure that given the input it create the expected output (in this case in files).

**Comment by DT-Priyanka** [comment]
@shubham-pathak22  added tests to validate input data. Please review and merge.

File: tutorials/parser/README.md [hunk 6]
```diff
@@ -1,4 +1,27 @@
...
+
+* **Csv Parser App**
+
+This application showcases how to use [CsvParser](https://datatorrent.com/docs/apidocs/com/datatorrent/contrib/parser/CsvParser.html) from [Apex Malhar](https://github.com/apache/apex-malhar) library. The CsvParser converts your delimited data to a key-value pair map or concrete java class also know as [POJO](https://en.wikipedia.org/wiki/Plain_Old_Java_Object). The parser emits key-value pair map on *parsedOutput* port. It emits POJO on *out* and error records on *err* port.
+
+Follow these steps to run this application:
+
+
+**Step 1**: Update/add these properties in the file `src/main/resources/META-INF/properties-csvParseApplication.xml`:
+
```
Comment (shubham-pathak22) [inline tutorials/parser/README.md]: Why step 1 is necessary ?? I can see they are already there in the properties file. Can't i just build the app the run ?   Or do you want to say which properties users need to specify to use csv parser ? If so, its not clear .

**Pull request merged**